Implement TimerService with an intrusive pairing heap.

goffrie opened the pull request:
This makes the timer more efficient when it has many nodes (O(1) rather
than O(n) insertion, and amortized O(log n) removal).

Matthias247 commented on src/intrusive_pairing_heap.rs:
```diff
@@ -0,0 +1,402 @@
+    struct DropBomb;
+    impl Drop for DropBomb {
+        fn drop(&mut self) {
+            panic!("Panicked while comparing");
```
Any benefit compared to aborting?

goffrie replied:
Is there an easy way to abort in `no_std`?

Matthias247 commented on src/intrusive_pairing_heap.rs:
```diff
@@ -0,0 +1,402 @@
+    debug_assert!(!safe_lesser(&child.as_ref().data, &parent.as_ref().data));
+    if let Some(mut old_first_child) = parent.as_mut().first_child.take() {
+        child.as_mut().next = Some(old_first_child);
+        old_first_child.as_mut().prev = Some(child);
```
How about a `debug_assert!(old_first_child.prev.is_none())` before?

Matthias247 commented on src/intrusive_pairing_heap.rs:
```diff
@@ -0,0 +1,402 @@
+    debug_assert!(left.as_ref().is_root());
+    debug_assert!(right.as_ref().is_root());
+    if safe_lesser(&right.as_ref().data, &left.as_ref().data) {
+        mem::swap(&mut left, &mut right);
```
This just swaps the pointers, and not their content, right?

Might be easier to understand if we just call `add_child(right, left)`  in this branch and use an `else` after it.

goffrie replied:
True.

Matthias247 commented on src/intrusive_pairing_heap.rs:
```diff
@@ -0,0 +1,402 @@
+unsafe fn last_child<T>(
+    mut first_child: NonNull<HeapNode<T>>,
+) -> NonNull<HeapNode<T>> {
+    while let Some(next) = first_child.as_ref().next {
```
Nit: I would have used an temporary variable, since keeping the `first_child` name is a bit weird.

Matthias247 replied:
This also raises the question why a node doesn't store another pointer to its last child, in order to avoid the iteration. Is the child list not that big, so that it doesn't matter and the memory savings are more important?

goffrie replied:
Hm, I guess it would need to be benchmarked. Whenever this method is called, the child list must be traversed in its entirety anyway, so it's linear-time anyway; and in general, we "hope" (amortized analysis) that the child list is fairly short; but it's true that this can be arbitrarily long.

The biggest reason why I chose this way for now is that maintaining the last-child pointer is extra effort :P

Matthias247 commented on src/intrusive_pairing_heap.rs:
```diff
@@ -0,0 +1,402 @@
+) -> Option<NonNull<HeapNode<T>>> {
+    let mut prev = node.as_mut().prev.take()?;
+    debug_assert_eq!(prev.as_ref().next, Some(node));
+    prev.as_mut().next = None;
```
What is with the previous nodes before the unlinked one? Do we need to adjust pointers?

```
prev.prev.next = Some(next);
```` 

and what about parent/child pointers on the unlinked node?

goffrie replied:
`prev.prev` shouldn't be touched; the idea is that given a list of the form
```
A <-> B <-> C
```
calling `unlink_prev` on `C` should mutate the list to become
```
A <-> B      C
```
and return `B`.

Parent/child pointers aren't touched by this function.

I'll add a comment.

Matthias247 commented on src/intrusive_pairing_heap.rs:
```diff
@@ -0,0 +1,402 @@
+            Insert(u8),
+            Remove(u8),
+        }
+        fn generate_schedules(
```
Nit: An empty line before/after function/type declarations makes things easier to read

Matthias247 commented on src/timer/timer.rs:
```diff
@@ -283,12 +287,10 @@ impl<MutexType: RawMutex> GenericTimerService<MutexType> {
     /// Returns a deadline based on the current timestamp plus the given Duration
     fn deadline_from_now(&self, duration: Duration) -> u64 {
         // TODO: Make this more efficient and with better overflow checking
```
I think we can remove this comment with that change. I didn't think about `saturating_add` back then.

goffrie commented:
Sorry for the delayed response, I've addressed your comments (hopefully). 

> I'll also open up the question on whether this is now the best implementation for a timer, or whether e.g. a wheel or a skiplist with skiplanes that are defined in terms of time-units (e.g. N ms) would be preferable. I don't know, but this is certainly already better than a pure linked list.

Great question. I have to admit that I mostly implemented this as a curiosity ;) It's certainly likely that other data structures could perform better in practice. However, some kind of node-based heap (i.e. not a `BinaryHeap`) does seem to be a good fit since it is very to adapt to the intrusive pattern, and I also like the fact that it doesn't enforce any particular granularity.

Matthias247 commented:
Thanks for this neat improvement! I don't have any use-case for the timer myself. But I'm sure if someone will pick it up it will be helpful for them.
